### PR TITLE
Add Timeit function for timecost calculations.

### PR DESCRIPTION
**Pull request by zongzw**
**Description**: 

This PR is used to calculate function level execution time.

Effects:

```
2022/01/24 06:58:49 [DEBUG] /root/workspace/src/github.com/F5Networks/k8s-bigip-ctlr/pkg/agent/as3/as3Resource.go 28 github.com/F5Networks/k8s-bigip-ctlr/pkg/agent/as3.(*AS3Manager).prepareAS3ResourceConfig (0 ms): 
2022/01/24 06:58:53 [INFO] github.com/F5Networks/k8s-bigip-ctlr/pkg/agent/as3.(*AS3Manager).postAS3Declaration (3961 ms): 

```

**Changes Proposed in PR**:

Refer to `Timeit` function.

**Fixes**: 

No issue ID but performance related.

## General Checklist
- [x] Updated Added functionality/ bug fix in release notes
- [x] Added examples for new feature
- [x] Updated the documentation
- [x] Smoke testing completed

## CRD Checklist
- [x] Updated required CR schema

### CODE REVIEW

**Comment by trinaths** [comment]
Created [CONTCNTR-3201] for internal tracking.

**Review by arzzon — CHANGES_REQUESTED** [review]

File: Makefile [hunk 1]
```diff
@@ -7,7 +7,8 @@ GOBIN    = $(GOPATH)/bin/$(GOOS)-$(GOARCH)
 
 NEXT_VERSION := $(shell ./build-tools/version-tool version)
 export BUILD_VERSION := $(if $(BUILD_VERSION),$(BUILD_VERSION),$(NEXT_VERSION))
-export BUILD_INFO := $(shell ./build-tools/version-tool build-info)
+NEXT_INFO := $(shell ./build-tools/version-tool build-info)
+export BUILD_INFO := $(if $(BUILD_INFO),$(BUILD_INFO),$(NEXT_INFO))
```
[the remark this answers is not in the document]
Comment (zongzw) [reply Makefile]: Removed from this PR, it was placed as a separate PR: #2373

File: pkg/prometheus/prometheus.go [hunk 2]
```diff
@@ -31,11 +29,19 @@ var CurrentErrors = prometheus.NewGaugeVec(
...
+		Name: "bigip_monitored_function_time_cost",
+		Help: "Recording function execution time cost for performance tunning",
+	},
+	[]string{"function"},
+)
+
 // further metrics? todo think about
 // RegisterMetrics registers all Prometheus metrics defined above
 func RegisterMetrics() {
-	log.Info("[CORE] Registered BigIP Metrics")
```
[the remark this answers is not in the document]
Comment (zongzw) [reply pkg/prometheus/prometheus.go]: The PR is stale, I just update it to date.

[the remark this answers is not in the document]
Comment (zongzw) [reply pkg/prometheus/prometheus.go]: The PR is updated to date. Please help to review. Thanks.

File: pkg/agent/as3/as3Common.go [hunk 3]
```diff
@@ -60,6 +60,8 @@ func (am *AS3Manager) validateAS3Template(template string) bool {
 }
 
 func (am *AS3Manager) processResourcesForAS3(sharedApp as3Application) {
+	defer log.Timeit("info")("")
```
Comment (arzzon) [inline pkg/agent/as3/as3Common.go:63]: Please change info to debug as at info level we shouldn't log function execution time.

File: pkg/agent/as3/as3Manager.go [hunk 4]
```diff
@@ -213,6 +213,7 @@ func updateTenantMap(tempAS3Config AS3Config) AS3Config {
 }
 
 func (am *AS3Manager) postAS3Declaration(rsReq ResourceRequest) (bool, string) {
+	defer log.Timeit("info")("")
```
Comment (arzzon) [inline pkg/agent/as3/as3Manager.go:216]: Please change info to debug here as well. Rest all look good.

**Review by arzzon — CHANGES_REQUESTED** [review]

File: pkg/agent/as3/as3Common.go (hunk 3 above)
Comment (zongzw) [reply pkg/agent/as3/as3Common.go]: done

File: pkg/agent/as3/as3Manager.go (hunk 4 above)
Comment (zongzw) [reply pkg/agent/as3/as3Manager.go]: Done

**Review by arzzon — APPROVED** [review]
LGTM

**Comment by arzzon** [comment]
@zongzw Could you please rebase it?

**Comment by zongzw** [comment]
> 

done.

**Pull request merged**